### PR TITLE
Add rmm::prefetch() and  DeviceBuffer.prefetch()

### PR DESCRIPTION
**Pull request by harrism**
## Description
This adds two `rmm::prefetch()` functions in C++.
1. `rmm::prefetch(void *ptr, size_t bytes, device, stream)`
2. `rmm::prefetch<T>(cuda::std::span<T> data, device, stream)`

Item 2 enables prefetching the containers that RMM provides (`device_uvector`, `device_scalar`) that support conversion to `cuda::std::span`. In order to enable that, `device_scalar::size()` is added.

Note that `device_buffer`s must be prefetched using item 1 because you can't create a `span<void>`.

In Python, this adds `DeviceBuffer.prefetch()` because that's really the only RMM Python data type to prefetch. There is *one* Cython use of `device_uvector` in cuDF `join` that we might need to add prefetch support for later.

`prefetch` is a no-op on non-managed memory. Rather than querying the type of memory, it just catches `cudaErrorInvalidValue` from `cudaMemPrefetchAsync`.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: include/rmm/device_buffer.hpp
Comment (madsbk) [inline include/rmm/device_buffer.hpp:362]:
```suggestion

    This is a no-op if the buffer isn't managed memory.
    
    @throw rmm::cuda_error if the prefetch fails.
```

**Comment by jrhemstad** [comment]
I don't think this should be a member function. The fact that the member function is only useful for specific upstream resource types is a big code smell to me.

Instead, I think this should be a freestanding function like:
```
template <typename T>
auto prefetch(T const* ptr, size_t n, device, stream);
```

**Comment by harrism** [comment]
> I don't think this should be a member function. The fact that the member function is only useful for specific upstream resource types is a big code smell to me. Instead, I think this should be a freestanding function 

I actually had the same thought over the weekend. But there are a few other factors to consider.

1. In RMM Python, the only thing one has to prefetch is a `DeviceBuffer`. And pointers are non-Pythonic. So I think it makes sense to associate prefetch with `DeviceBuffer` in the python API, either as a method or a free function that expects a DeviceBuffer.
2. I do think we want to put unconditional calls to `prefetch` in algorithm code. These should be functional no-ops when the memory is not managed.  (A `migrateable` or similar property for `cuda::mr` would help with that...)
3. What do you think about versions for containers and iterators?  I think only providing a pointer interface is error-prone, and makes it harder than necessary to prefetch a range from a container, or a whole container.

```
template <typename Iterator>
auto prefetch(Iterator begin, Iterator end, device, stream)
{
  constexpr auto elt_size = sizeof(std::iterator_traits<Iterator>::value_type);
  return prefetch(begin, std::distance(begin, end) * elt_size, device, stream);
}

template <typename Container>
auto prefetch(Container const& c, device, stream) 
{ 
  return prefetch(c.begin(), c.end(), device, stream);
}
```

File: python/rmm/docs/guide.md
Comment (wence-) [inline python/rmm/docs/guide.md:74]:
```suggestion
CUDA Unified Memory, also known as managed memory, can be allocated using an
`rmm.mr.ManagedMemoryResource` explicitly, or by calling `rmm.reinitialize` with
`managed_memory=True`.
```

Comment (wence-) [reply python/rmm/docs/guide.md:74]: *question*: Shall we link "CUDA Unified Memory` anywhere? Perhaps https://developer.nvidia.com/blog/unified-memory-cuda-beginners/ ?

File: python/rmm/docs/guide.md
Comment (wence-) [inline python/rmm/docs/guide.md:76]:
```suggestion
A `DeviceBuffer` backed by managed memory may be prefetched to a specified
```

File: python/rmm/docs/guide.md
Comment (wence-) [inline python/rmm/docs/guide.md:87]:
```suggestion
The above example prefetches the `DeviceBuffer` memory to the current CUDA device
on the stream that the `DeviceBuffer` last used (e.g. at construction). The
```

File: python/rmm/docs/guide.md
Comment (wence-) [inline python/rmm/docs/guide.md:96]:
```suggestion
>>> buf = rmm.DeviceBuffer(size=100, stream=stream)
>>> buf.prefetch(device=3, stream=stream) # prefetch to device on stream.
```

We can't have positional arguments after keyword arguments in Python.

File: python/rmm/docs/guide.md
Comment (wence-) [inline python/rmm/docs/guide.md:99]:
```suggestion
`DeviceBuffer.prefetch()` is a no-op if the `DeviceBuffer` is not backed
```

File: include/rmm/prefetch.hpp
Comment (wence-) [inline include/rmm/prefetch.hpp:66]: *note (nothing to do here)*: A slight annoyance that without concepts I think there's no way to avoid writing the template parameter at the call site, but if we had C++20 then:
```c++
template <typename T>
requires requires(T&& s) { cuda::std::span(std::forward<T>(s)); }
void prefetch(T&& data, rmm::cuda_device_id device, rmm::cuda_stream_view stream)
{
  cuda::std::span sp(std::forward<T>(data));
  prefetch(sp.data(), sp.size_bytes(), device, stream);
}
```
Would allow one to write:
```c++
rmm::device_vector<int> buf{...};
prefetch(buf, ...);
```
Without the need for the template parameter to `prefetch`, I think.

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (wence-) [inline python/rmm/rmm/_lib/device_buffer.pyx:141]: *question*: In the C++ interface, the device and stream are required. Do we think it makes sense for them to be required in Python too? Or is the likely usage almost always "fetch on the current device and the stream associated with the buffer"?

**Review by wence- — APPROVED** [review]
Some small docs fixes and one question about optional arguments in the python interface.

**Review by rongou — APPROVED** [review]
I think with HMM/ATS, system allocated memory (malloc/free) can also be prefetched, but we can cross that bridge when we come to it.

File: python/rmm/docs/guide.md
Comment (harrism) [reply python/rmm/docs/guide.md:76]: Sorry, I was thinking doxygen where known names are automatically highlighted. I noticed there are other uses of DeviceBuffer in this file not quoted, will have a look.

File: python/rmm/docs/guide.md
Comment (harrism) [reply python/rmm/docs/guide.md:96]: Right, I knew that. 🤦

File: include/rmm/prefetch.hpp
Comment (harrism) [reply include/rmm/prefetch.hpp:66]: Yes, the current required interface to support `span` is annoying.

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:141]: I was unsure about this. In RMM containers, we make sure the stream is always explicit for stream safety reasons. It's a stream-ordered container so we felt that explicit streams are less error prone.

But in the Python API we have a lot of default values for the stream parameter, so I copied that style here. We may want to revisit this discussion across the whole Python API rather than making a specific decision here that might mismatch current practice?

File: include/rmm/prefetch.hpp
Comment (harrism) [reply include/rmm/prefetch.hpp:66]: I need to learn more about concepts to understand what your code is saying.

**Comment by vyasr** [comment]
Apologies, I had intended to review this but haven't quite made the time for it yet and things have moved along quite quickly. I'll attempt to review tomorrow, but don't let my missing review stop you from merging if you think this is ready.

File: python/rmm/docs/guide.md
Comment (harrism) [reply python/rmm/docs/guide.md]: Add this link, and one for HMM.

**Comment by harrism** [comment]
> I think with HMM/ATS, system allocated memory (malloc/free) can also be prefetched, but we can cross that bridge when we come to it.

@rongou I crossed the bridge: made the docs a bit less specific to managed memory based on your suggestion, added a mention of / link to HMM. I couldn't find a good link for ATS though.

File: python/rmm/docs/guide.md
Comment (harrism) [reply python/rmm/docs/guide.md]: I found some more instances of unquoted MemoryResource and DeviceBuffer (notably when plural) and fixed them.

**Comment by harrism** [comment]
> Apologies, I had intended to review this but haven't quite made the time for it yet and things have moved along quite quickly. I'll attempt to review tomorrow, but don't let my missing review stop you from merging if you think this is ready.

No worries, still need a cmake codeowners approval anyway, and I would like @jrhemstad to approve since he complained about the `device_buffer` method approach I originally took. Also @miscco for the `span` stuff.

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (vyasr) [inline python/rmm/rmm/_lib/device_buffer.pyx:368]: Having a separate function for this seems excessive when it could just be inlined into `prefetch` above. To keep things concise you can set the first argument with a ternary in Python and Cython shouldn't have a problem with it; `prefetch(cuda_device_id(device) if device is not None else get_current_cuda_device()...)`

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (vyasr) [inline python/rmm/rmm/_lib/device_buffer.pyx:141]: It feels slightly odd to have a free function in C++ but a method in Python. I guess this is more practically useful though? Also we can't have overloads in Python like you can in C++.

File: tests/prefetch_tests.cpp
Comment (vyasr) [inline tests/prefetch_tests.cpp:57]: Could we verify that the data is actually resident in GPU memory using [`cudaMemRangeGetAttribute`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g8048f6ea5ad77917444567656c140c5a) or similar?

**Review by vyasr — COMMENTED** [review]
I ended up just going ahead and reviewing :)

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:141]: There are no pointers in Python. RMM Python represents data allocations using DeviceBuffer. So it makes sense to only prefetch DeviceBuffers.

It could be a free function though, and then it could introspect the type to decide how to prefetch it (runtime overload, effectively).

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:368]: I can't get it to cython compile with 

```
prefetch(self.c_obj.get()[0].data(),
                self.c_obj.get()[0].size(),
                dev,
                stream.view())
```

```
/home/coder/rmm/python/rmm/rmm/_lib/device_buffer.pyx:161:27: Cannot convert Python object to 'cuda_stream_view'
```

File: tests/prefetch_tests.cpp
Comment (harrism) [reply tests/prefetch_tests.cpp:77]: Wow, I did not know about this function. Great idea! Added.

File: include/rmm/prefetch.hpp
Comment (wence-) [reply include/rmm/prefetch.hpp:66]: > doesn't calling this function with an rmm::device_buffer automatically convert to a span? 

Yes, but template argument deduction doesn't work. So you have to write:

```c++
rmm::device_buffer<int> buf{...};
prefetch<int>(buf, ...);
```

Rather than what you feel like should work:
```c++
rmm::device_buffer<int> buf{...};
prefetch(buf, ...);
```

And yes, that requires concept is basically the concepts version of your SFINAE constraint. It says, I'll take any && reference as long as it is span constructible.

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (vyasr) [reply python/rmm/rmm/_lib/device_buffer.pyx:368]: It's complaining becuase in the function signature for this function `c_prefetch` the fourth argument is typed as a `Stream`, so when you call the underlying C function `prefetch` it knows the type of that object. Conversely, when you look at the Python `prefetch` implementation above, the `stream` object is not typed. You can fix that by doing a `cdef Stream stream =...` and it should work.

File: include/rmm/prefetch.hpp
Comment (wence-) [reply include/rmm/prefetch.hpp:66]: I mistyped, sub device_vector<int> for device_buffer

File: include/rmm/prefetch.hpp
Comment (jrhemstad) [inline include/rmm/prefetch.hpp:67]: This function shouldn't be mutating any of the data. 

```suggestion
void prefetch(cuda::std::span<T const> data, rmm::cuda_device_id device, rmm::cuda_stream_view stream)
```

File: include/rmm/prefetch.hpp
Comment (jrhemstad) [inline include/rmm/prefetch.hpp:47]: If `ptr` is typed, then `size` shouldn't be _bytes_, it should be _elements_.

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx]: Thanks!  Fixed.

File: python/rmm/rmm/tests/test_rmm.py
Comment (bdice) [inline python/rmm/rmm/tests/test_rmm.py:293]: You might be able to test that a prefetch call was issued with [`cudaMemRangeGetAttribute (void* data, size_t dataSize, cudaMemRangeAttribute attribute, const void* devPtr, size_t count)`, with attribute `cudaMemRangeAttributeLastPrefetchLocation`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g8048f6ea5ad77917444567656c140c5a). It ought to be possible to call this via `cuda-python` ([API docs](https://nvidia.github.io/cuda-python/module/cuda.html#cuda.cuda.cuMemRangeGetAttribute)).

Also:
> Note that this simply returns the last location that the applicaton requested to prefetch the memory range to. It gives no indication as to whether the prefetch operation to that location has completed or even begun.

You wouldn't know if the prefetch completed or not, but you could verify that the prefetch request was issued.

Comment (bdice) [reply python/rmm/rmm/tests/test_rmm.py:293]: ...of course, as soon as I scrolled down, I see you did this exact thing in the C++ tests, at Vyas's request. It would be nice to have a corresponding Python API test, since it should be quick to write with cuda-python.

File: include/rmm/prefetch.hpp
Comment (harrism) [reply include/rmm/prefetch.hpp:47]: I'm switching it back to `void const*` because then we can use span::size_bytes() in the span function.  Someone suggested the T* version during discussions but I can't remember who or why. If there is a good reason, I'm all ears.

File: python/rmm/rmm/tests/test_rmm.py
Comment (harrism) [reply python/rmm/rmm/tests/test_rmm.py:293]: OK...

Comment (harrism) [reply python/rmm/rmm/tests/test_rmm.py:307]: CUDA Python results in very ugly code due to its non-pythonic error handling. But I've done what you asked...

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (leofang) [inline python/rmm/rmm/_lib/device_buffer.pyx:151]: Q: Maybe this is just a docstring question, but what's the Python story for prefetching on CPU (`cudaCpuDeviceId`)? On the C++ side, we can initialize `rmm::cuda_device_id` to `cudaCpuDeviceId` (which is -1) without problem. 

This is not blocking, just curious about the CPU story since HMM/ATS was brought up in the reviews above.

Comment (jrhemstad) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: >  but what's the Python story for prefetching on CPU (cudaCpuDeviceId)? On the C++ side

NUMA nodes.

**Review by jrhemstad — APPROVED** [review]

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (leofang) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: Mind a few more words?

Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: I don't know how to do the NUMA nodes @jrhemstad mentions, but CUDA Python docs mention that it supports passing `cudaCpuDeviceId` just like in C++. https://nvidia.github.io/cuda-python/module/cudart.html#cuda.cudart.cudaMemPrefetchAsync

Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: Ah, now I see there is a `cudaMemPrefetchAsync_v2` of which I was previously unaware that takes more options for `dstLocation`. Notice that this version is undocumented for CUDA Python: https://nvidia.github.io/cuda-python/module/cudart.html#cuda.cudart.cudaMemPrefetchAsync_v2

**Review by miscco — APPROVED** [review]

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (leofang) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: It seems `cudaMemPrefetchAsync_v2` comes without any docstring, so it is undocumented everywhere, let me check internally.

Regardless, I would like to have @rapidsai/rmm-python-codeowners opinions here. I don't think you get my question, so let me rephrase: Do we expect Pythonusers to know that they need to pass -1 here for prefetching CPU? Why? Is it mentioned anywhere that this is the way to go?

Comment (leofang) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: Whether this is the right UX is subject to discussion, but the minimal work needed is something along this line:
```suggestion
        device : optional
            The CUDA device to which to prefetch the memory for this buffer.
            Defaults to the current CUDA device. To prefetch on CPU, pass ``-1``.
```

Comment (wence-) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: NACK: "To prefetch on CPU" makes no sense as a sentence, this code is executing on the CPU.

If we want this in the docstring at all, explicitly mention the (exposed by cuda-python) `cudaCpuDeviceId` constant

> To prefetch to the CPU, pass the ``cudaCpuDeviceId`` from cudart as the device.

Comment (leofang) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: OK. So we're ok to pass a pre-defined constant here. Then I suggest 
1. Add `"cuda-python": ("https://nvidia.github.io/cuda-python/", None),` to this dict:
https://github.com/rapidsai/rmm/blob/f47ce3f0d46848cd9d5844d499bf150dd14d823a/python/rmm/docs/conf.py#L209-L212
2. Update the docstring per Lawrence's suggestion, with cross-referencing to CUDA Python docs:
```suggestion
        device : optional
            The CUDA device to which to prefetch the memory for this buffer.
            Defaults to the current CUDA device. To prefetch to the CPU, pass
            `~cuda.cudart.cudaCpuDeviceId` as the device.
```

**Comment by wence-** [comment]
Devcontainer build fails are due to (I think) not having https://github.com/NVIDIA/cccl/pull/1836

**Review by vyasr — APPROVED** [review]

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: I was looking at this yesterday. Once 12.2 is our minimum version, we can switch to `_v2` and and modify the prefetching API to accept a locationType parameter. 

Or, I could make an API that does that now, but will only work for NUMA prefetching on 12.2 and later.

Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: What is the `~` in your code?

Comment (leofang) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: `~` would omit the full module qualifier so as to let Sphinx render it to just `cudaCpuDeviceId` (but with the correct hyperlink, for which a full qualifier is needed to compute).

**Comment by harrism** [comment]
> Devcontainer build fails are due to (I think) not having [NVIDIA/cccl#1836](https://github.com/NVIDIA/cccl/pull/1836)

I don't understand. This was working previously. I think this broke with the update to CCCL 2.5!

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (harrism) [reply python/rmm/rmm/_lib/device_buffer.pyx:151]: Thanks, updated docstring.

**Comment by leofang** [comment]
It seems the doc-build pipeline failed but there's no log. Is it possible to retrigger it?

File: python/rmm/rmm/_lib/device_buffer.pyx
Comment (leofang) [inline python/rmm/rmm/_lib/device_buffer.pyx:152]: It seems the doc build did not render it right. Possibly because the default role in RMM is "cpp" instead of "py". If so the fix would be
```
:py:`~cuda.cudart.cudaCpuDeviceId`
```

Comment (leofang) [reply python/rmm/rmm/_lib/device_buffer.pyx:152]: I am certain that the correct reference is there. It can be checked as follows:
```shell
$ python -m sphinx.ext.intersphinx https://nvidia.github.io/cuda-python/objects.inv | grep "cuda.cudart.cudaCpuDeviceId"
    cuda.cudart.cudaCpuDeviceId                                                      : module/cudart.html#cuda.cudart.cudaCpuDeviceId
```

Comment (jakirkham) [reply python/rmm/rmm/_lib/device_buffer.pyx:152]: Did this get fixed or should this be raised in a new issue?

Comment (bdice) [reply python/rmm/rmm/_lib/device_buffer.pyx:152]: I don’t think this was seen after merging. Issue filed: https://github.com/rapidsai/rmm/issues/1635#issue-2442423568

Comment (bdice) [reply python/rmm/rmm/_lib/device_buffer.pyx:152]: Fix: https://github.com/rapidsai/rmm/pull/1636#issue-2442432052

Comment (jakirkham) [reply python/rmm/rmm/_lib/device_buffer.pyx:152]: Thanks Bradley! 🙏

**Pull request merged**
